energy-integration/fees-collector/src/fees_accumulation.rs

```diff
@@ -21,32 +20,47 @@ pub trait FeesAccumulationModule:
             "Only known contracts can deposit"
         );
 
-        let (payment_token, payment_amount) = self.call_value().single_fungible_esdt();
+        let payment = self.call_value().single_esdt();
         require!(
-            self.known_tokens().contains(&payment_token),
+            self.known_tokens().contains(&payment.token_identifier),
             "Invalid payment token"
         );
 
         let current_week = self.get_current_week();
-        self.accumulated_fees(current_week, &payment_token)
-            .update(|amt| *amt += &payment_amount);
+        if payment.token_nonce == 0 {
+            self.accumulated_fees(current_week, &payment.token_identifier)
+                .update(|amt| *amt += &payment.amount);
+        } else {
+            self.accumulated_locked_fees(current_week, &payment.token_identifier)
+                .update(|locked_fees| locked_fees.push(payment.clone()));
+        }
 
-        self.emit_deposit_swap_fees_event(caller, current_week, payment_token, payment_amount);
+        self.emit_deposit_swap_fees_event(
+            caller,
+            current_week,
+            payment.token_identifier,
+            payment.amount,
+        );
     }
 
     fn collect_accumulated_fees_for_week(
         &self,
         week: Week,
-    ) -> ManagedVec<TokenAmountPair<Self::Api>> {
+    ) -> ManagedVec<EsdtTokenPayment<Self::Api>> {
         let mut results = ManagedVec::new();
         let all_tokens = self.all_tokens().get();
         for token in &all_tokens {
             let opt_accumulated_fees = self.get_and_clear_acccumulated_fees(week, &token);
             if let Some(accumulated_fees) = opt_accumulated_fees {
-                results.push(TokenAmountPair::new(token, accumulated_fees));
+                results.push(EsdtTokenPayment::new(token.clone(), 0, accumulated_fees));
             }
-        }
 
+            let opt_accumulated_locked_fees =
```

Comment:
I think you can make a separate FOR for this and you need token and nonce.

Reply:
Why iterate through all tokens twice? 
`opt_accumulated_locked_fees`, if not empty, will be an EsdtTokenPayment and will contain the nonce.